File: workspace/merger.go
package workspace

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v3"
)

var positionKeys = map[string]bool{
	"position_x": true,
	"position_y": true,
}

// MergeWorkspace merges changes from a new workspace (from server) into the current on-disk state.
// It uses yaml.Node to preserve comments and formatting.
// lastSyncMeta is the metadata from the .tld.lock file (state at last pull/apply).
// currentMeta is the metadata loaded from local YAML files (current state on disk).
func MergeWorkspace(dir string, newWS *Workspace, lastSyncMeta *Meta, currentMeta *Meta) error {
	if useElementWorkspaceFiles(newWS) {
		elementMetaSections := []metadataSection{{name: "_meta_elements", values: newWS.Meta.Elements}, {name: "_meta_views", values: newWS.Meta.Views}}
		if err := mergeYAMLMapWithMetadataSections(
			filepath.Join(dir, "elements.yaml"),
			newWS.Elements,
			combinedElementMetadata(newWS.Meta),
			combinedElementMetadata(lastSyncMeta),
			combinedElementMetadata(currentMeta),
			elementMetaSections,
		); err != nil {
			return fmt.Errorf("merge elements: %w", err)
		}

		if err := mergeYAMLMapWithMetadataSections(
			filepath.Join(dir, "connectors.yaml"),
			newWS.Connectors,
			newWS.Meta.Connectors,
			lastSyncMeta.Connectors,
			currentMeta.Connectors,
			[]metadataSection{{name: "_meta_connectors", values: newWS.Meta.Connectors}},
		); err != nil {
			return fmt.Errorf("merge connectors: %w", err)
		}

		if err := cleanupLegacyWorkspaceFiles(dir); err != nil {
			return fmt.Errorf("cleanup legacy workspace files: %w", err)
		}
		return nil
	}

	return nil
}

func mergeYAMLMapWithConflicts(path string, serverItems any, serverMeta map[string]*ResourceMetadata, lastSyncMeta map[string]*ResourceMetadata, currentMeta map[string]*ResourceMetadata) error {
	return mergeYAMLMapWithMetadataSections(path, serverItems, serverMeta, lastSyncMeta, currentMeta, []metadataSection{{name: "_meta", values: serverMeta}})
}

func mergeYAMLMapWithMetadataSections(path string, serverItems any, serverMeta map[string]*ResourceMetadata, lastSyncMeta map[string]*ResourceMetadata, currentMeta map[string]*ResourceMetadata, sections []metadataSection) error {
	// Load existing file into a Node
	var root yaml.Node
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return WriteFullYAMLMapSections(path, serverItems, sections)
		}
		return fmt.Errorf("read %s: %w", path, err)
	}

	if err := yaml.Unmarshal(data, &root); err != nil {
		return fmt.Errorf("unmarshal %s: %w", path, err)
	}

	if root.Kind != yaml.DocumentNode || len(root.Content) == 0 || root.Content[0].Kind != yaml.MappingNode {
		return WriteFullYAMLMapSections(path, serverItems, sections)
	}

	mapping := root.Content[0]

	// Convert serverItems to a map for easy lookup
	serverItemsData, _ := yaml.Marshal(serverItems)
	var serverItemsMap map[string]any
	_ = yaml.Unmarshal(serverItemsData, &serverItemsMap)

	seenKeys := make(map[string]bool)
	var newContent []*yaml.Node

	// Iterate through existing mapping (local state)
	for i := 0; i < len(mapping.Content); i += 2 {
		keyNode := mapping.Content[i]
		valNode := mapping.Content[i+1]
		key := keyNode.Value

		if isMetadataSectionKey(key, sections) {
			continue
		}

		serverItem, onServer := serverItemsMap[key]
		sMeta := serverMeta[key]
		lMeta := lastSyncMeta[key]
		cMeta := currentMeta[key]

		if onServer {
			// Resource exists locally and on server
			seenKeys[key] = true

			localChanged := lMeta != nil && cMeta != nil && cMeta.UpdatedAt.After(lMeta.UpdatedAt)
			serverChanged := lMeta != nil && sMeta != nil && sMeta.UpdatedAt.After(lMeta.UpdatedAt)

			if localChanged && serverChanged {
				mergedNode, hasConflict, mergeErr := mergeResourceValueNode(valNode, serverItem, key)
				if mergeErr != nil {
					return fmt.Errorf("merge %s[%s]: %w", filepath.Base(path), key, mergeErr)
				}
				if hasConflict {
					for _, section := range sections {
						if section.values[key] != nil {
							section.values[key].Conflict = true
							break
						}
					}
				} else if sMeta != nil {
					sMeta.Conflict = false
				}
				newContent = append(newContent, keyNode, mergedNode)
			} else if serverChanged {
				// Server changed, local did not: Update local with server content.
				var newValNode yaml.Node
				_ = newValNode.Encode(serverItem)
				newContent = append(newContent, keyNode, &newValNode)
			} else {
				// No changes or only local changes: Keep local.
				newContent = append(newContent, keyNode, valNode)
			}
		} else {
			// Resource exists locally but not on server.
			// Was it deleted on server or created locally?
			if lMeta == nil {
				// Was never on server, so it's a new local resource. Keep it.
				newContent = append(newContent, keyNode, valNode)
			}
			// Else: Was on server before (last sync), so it was deleted on server.
			// Remove locally too by NOT adding to newContent.
		}
	}

	// Add new keys from server that weren't in the local file
	for key, val := range serverItemsMap {
		if !seenKeys[key] {
			var keyNode yaml.Node
			_ = keyNode.Encode(key)
			var valNode yaml.Node
			_ = valNode.Encode(val)
			newContent = append(newContent, &keyNode, &valNode)
		}
	}

	for _, section := range sections {
		if len(section.values) == 0 {
			continue
		}
		var metaKeyNode yaml.Node
		_ = metaKeyNode.Encode(section.name)
		metaValNode, err := EncodeMeta(section.values)
		if err != nil {
			return err
		}
		newContent = append(newContent, &metaKeyNode, metaValNode)
	}

	mapping.Content = newContent

	// Write back
	normalizeYAMLStyle(&root)
	f, err := os.Create(path)
	if err != nil {
		return fmt.Errorf("create %s: %w", path, err)
	}
	defer func() { _ = f.Close() }()

	enc := yaml.NewEncoder(f)
	enc.SetIndent(2)
	if err := enc.Encode(&root); err != nil {
		return fmt.Errorf("encode %s: %w", path, err)
	}
	return nil
}

func isMetadataSectionKey(key string, sections []metadataSection) bool {
	for _, section := range sections {
		if key == section.name {
			return true
		}
	}
	return false
}

func combinedElementMetadata(meta *Meta) map[string]*ResourceMetadata {
	combined := make(map[string]*ResourceMetadata)
	if meta == nil {
		return combined
	}
	for ref, resourceMeta := range meta.Elements {
		if resourceMeta == nil {
			continue
		}
		copyMeta := *resourceMeta
		combined[ref] = &copyMeta
	}
	for ref, resourceMeta := range meta.Views {
		if resourceMeta == nil {
			continue
		}
		existing := combined[ref]
		if existing == nil {
			copyMeta := *resourceMeta
			combined[ref] = &copyMeta
			continue
		}
		if resourceMeta.UpdatedAt.After(existing.UpdatedAt) {
			existing.UpdatedAt = resourceMeta.UpdatedAt
		}
		if existing.ID == 0 {
			existing.ID = resourceMeta.ID
		}
	}
	return combined
}

func mergeResourceValueNode(localNode *yaml.Node, serverItem any, ref string) (*yaml.Node, bool, error) {
	serverNode, err := encodeYAMLValueNode(serverItem)
	if err != nil {
		return nil, false, err
	}
	merged, hasConflict := mergeNodeValues(localNode, serverNode, "")
	if hasConflict && merged.HeadComment == "" {
		merged.HeadComment = fmt.Sprintf("CONFLICT: %q was modified both locally and on the server.\nResolve the marked values, then run `tld apply`.", ref)
	}
	return merged, hasConflict, nil
}

func mergeNodeValues(localNode, serverNode *yaml.Node, fieldName string) (*yaml.Node, bool) {
	if serverNode == nil {
		return cloneYAMLNode(localNode), false
	}
	if localNode == nil {
		return cloneYAMLNode(serverNode), false
	}
	if positionKeys[fieldName] {
		return cloneYAMLNode(serverNode), false
	}

	if localNode.Kind == yaml.MappingNode && serverNode.Kind == yaml.MappingNode {
		return mergeMappingNodes(localNode, serverNode)
	}
	if localNode.Kind == yaml.SequenceNode && serverNode.Kind == yaml.SequenceNode {
		if fieldName == "diagrams" || fieldName == "placements" {
			return mergePlacementSequences(localNode, serverNode, fieldName)
		}
		return cloneYAMLNode(serverNode), false
	}

	localValue := strings.TrimSpace(localNode.Value)
	serverValue := strings.TrimSpace(serverNode.Value)
	if localValue == serverValue {
		return cloneYAMLNode(localNode), false
	}
	if serverValue == "" && localValue != "" {
		return cloneYAMLNode(localNode), false
	}
	if localValue == "" && serverValue != "" {
		return cloneYAMLNode(serverNode), false
	}
	return conflictValueNode(fieldName, localValue, serverValue), true
}

func mergeMappingNodes(localNode, serverNode *yaml.Node) (*yaml.Node, bool) {
	result := &yaml.Node{Kind: yaml.MappingNode, Tag: "!!map"}
	hasConflict := false
	serverValues := make(map[string]*yaml.Node)
	serverOrder := make([]string, 0, len(serverNode.Content)/2)
	for i := 0; i+1 < len(serverNode.Content); i += 2 {
		serverValues[serverNode.Content[i].Value] = serverNode.Content[i+1]
		serverOrder = append(serverOrder, serverNode.Content[i].Value)
	}
	seen := make(map[string]bool)

	for i := 0; i+1 < len(localNode.Content); i += 2 {
		key := localNode.Content[i].Value
		localVal := localNode.Content[i+1]
		seen[key] = true
		mergedVal := cloneYAMLNode(localVal)
		fieldConflict := false
		if serverVal, ok := serverValues[key]; ok {
			mergedVal, fieldConflict = mergeNodeValues(localVal, serverVal, key)
		}
		result.Content = append(result.Content, cloneYAMLNode(localNode.Content[i]), mergedVal)
		hasConflict = hasConflict || fieldConflict
	}

	for _, key := range serverOrder {
		if seen[key] {
			continue
		}
		result.Content = append(result.Content,
			&yaml.Node{Kind: yaml.ScalarNode, Tag: "!!str", Value: key},
			cloneYAMLNode(serverValues[key]),
		)
	}
	return result, hasConflict
}

func mergePlacementSequences(localNode, serverNode *yaml.Node, fieldName string) (*yaml.Node, bool) {
	result := &yaml.Node{Kind: yaml.SequenceNode, Tag: "!!seq"}
	serverValues := make(map[string]*yaml.Node)
	serverOrder := make([]string, 0, len(serverNode.Content))
	for _, item := range serverNode.Content {
		if key := placementSequenceKey(item); key != "" {
			serverValues[key] = item
			serverOrder = append(serverOrder, key)
		}
	}
	seen := make(map[string]bool)
	hasConflict := false

	for _, localItem := range localNode.Content {
		key := placementSequenceKey(localItem)
		seen[key] = true
		if key != "" {
			if serverItem, ok := serverValues[key]; ok {
				mergedItem, conflict := mergeNodeValues(localItem, serverItem, fieldName)
				result.Content = append(result.Content, mergedItem)
				hasConflict = hasConflict || conflict
				continue
			}
		}
		result.Content = append(result.Content, cloneYAMLNode(localItem))
	}

	for _, key := range serverOrder {
		if seen[key] {
			continue
		}
		result.Content = append(result.Content, cloneYAMLNode(serverValues[key]))
	}
	return result, hasConflict
}

func placementSequenceKey(node *yaml.Node) string {
	if node == nil || node.Kind != yaml.MappingNode {
		return ""
	}
	for i := 0; i+1 < len(node.Content); i += 2 {
		switch node.Content[i].Value {
		case "diagram", "parent":
			return node.Content[i+1].Value
		}
	}
	return ""
}

func conflictValueNode(fieldName, localValue, serverValue string) *yaml.Node {
	comment := fmt.Sprintf("CONFLICT: '%s' was modified both locally and on the server.\nLocal:  %s\nServer: %s\nResolve by keeping one value, then run `tld apply`.", fieldName, localValue, serverValue)
	return &yaml.Node{
		Kind:        yaml.ScalarNode,
		Tag:         "!!str",
		Style:       yaml.LiteralStyle,
		HeadComment: comment,
		Value:       fmt.Sprintf("<<< LOCAL\n%s\n===\n%s\n>>> SERVER", localValue, serverValue),
	}
}

func cloneYAMLNode(node *yaml.Node) *yaml.Node {
	if node == nil {
		return nil
	}
	clone := *node
	if len(node.Content) > 0 {
		clone.Content = make([]*yaml.Node, len(node.Content))
		for i, child := range node.Content {
			clone.Content[i] = cloneYAMLNode(child)
		}
	}
	return &clone
}

File: src/workspace/merger.rs
use crate::error::TldError;
use crate::workspace::types::{Meta, ResourceMetadata, Workspace};
use std::collections::HashMap;
use std::fs;
use std::path::Path;
use yaml_rust2::{Yaml, YamlEmitter, YamlLoader};

/// Merges server state into local YAML files while preserving comments and formatting.
pub fn merge_workspace(
    wdir: &str,
    server_ws: &Workspace,
    last_sync: &Meta,
    current_meta: &Meta,
) -> Result<(), TldError> {
    let elements_path = Path::new(wdir).join("elements.yaml");
    let connectors_path = Path::new(wdir).join("connectors.yaml");

    let mut total_conflicts = 0;

    // Merge Elements
    if elements_path.exists() {
        total_conflicts += merge_file(
            &elements_path,
            &server_ws.elements,
            &server_ws
                .meta
                .as_ref()
                .map(|m| m.elements.clone())
                .unwrap_or_default(),
            &last_sync.elements,
            &current_meta.elements,
        )?;
    } else {
        crate::workspace::save(server_ws)?;
    }

    // Merge Connectors
    if connectors_path.exists() {
        total_conflicts += merge_file(
            &connectors_path,
            &server_ws.connectors,
            &server_ws
                .meta
                .as_ref()
                .map(|m| m.connectors.clone())
                .unwrap_or_default(),
            &last_sync.connectors,
            &current_meta.connectors,
        )?;
    }

    if total_conflicts > 0 {
        return Err(TldError::Generic(format!(
            "Merge complete with {} conflict(s). Please resolve them in elements.yaml/connectors.yaml and run 'tld apply' again.",
            total_conflicts
        )));
    }

    Ok(())
}

fn merge_file<T: serde::Serialize>(
    path: &Path,
    server_items: &HashMap<String, T>,
    server_meta: &HashMap<String, ResourceMetadata>,
    last_sync_meta: &HashMap<String, ResourceMetadata>,
    current_local_meta: &HashMap<String, ResourceMetadata>,
) -> Result<usize, TldError> {
    let content = fs::read_to_string(path).map_err(|e| TldError::Generic(e.to_string()))?;
    let docs = YamlLoader::load_from_str(&content).map_err(|e| TldError::Generic(e.to_string()))?;
    if docs.is_empty() {
        return Ok(0);
    }

    // We treat the first document as the mapping.
    let mut doc = docs[0].clone();
    let mut conflicts = 0;

    if let Yaml::Hash(ref mut h) = doc {
        let mut keys_to_remove = Vec::new();

        for (key, val) in h.iter_mut() {
            let key_str = match key.as_str() {
                Some(s) => s.to_string(),
                None => continue,
            };

            if key_str.starts_with("_meta") {
                continue;
            }

            let s_meta = server_meta.get(&key_str);
            let l_meta = last_sync_meta.get(&key_str);
            let c_meta = current_local_meta.get(&key_str);

            if let Some(server_item) = server_items.get(&key_str) {
                let local_changed = match (l_meta, c_meta) {
                    (Some(l), Some(c)) => c.updated_at > l.updated_at,
                    _ => false,
                };
                let server_changed = match (l_meta, s_meta) {
                    (Some(l), Some(s)) => s.updated_at > l.updated_at,
                    _ => false,
                };

                if local_changed && server_changed {
                    // CONFLICT
                    conflicts += 1;
                    let server_yaml_str = serde_yaml::to_string(server_item).unwrap_or_default();
                    let local_yaml_str = format!("{:?}", val); // Simplified local capture
                    let conflict_val = format!(
                        "<<< LOCAL\n{}\n===\n{}>>> SERVER",
                        local_yaml_str.trim(),
                        server_yaml_str.trim()
                    );
                    *val = Yaml::String(conflict_val);
                } else if server_changed {
                    // Update from server
                    let server_yaml_str = serde_yaml::to_string(server_item).unwrap_or_default();
                    let server_docs =
                        YamlLoader::load_from_str(&server_yaml_str).unwrap_or_default();
                    if !server_docs.is_empty() {
                        *val = server_docs[0].clone();
                    }
                }
            } else if l_meta.is_some() {
                // Deleted on server
                keys_to_remove.push(key.clone());
            }
        }

        for k in keys_to_remove {
            h.remove(&k);
        }

        // Add new items from server
        for (key, item) in server_items {
            let key_yaml = Yaml::String(key.clone());
            if h.get(&key_yaml).is_none() {
                let server_yaml_str = serde_yaml::to_string(item).unwrap_or_default();
                let server_docs = YamlLoader::load_from_str(&server_yaml_str).unwrap_or_default();
                if !server_docs.is_empty() {
                    h.insert(key_yaml, server_docs[0].clone());
                }
            }
        }
    }

    let mut out_str = String::new();
    let mut emitter = YamlEmitter::new(&mut out_str);
    emitter
        .dump(&doc)
        .map_err(|e| TldError::Generic(e.to_string()))?;
    fs::write(path, out_str).map_err(|e| TldError::Generic(e.to_string()))?;
    Ok(conflicts)
}
